GithubBrowserSample/app/src/main/java/com/android/example/github/repository/CoroutineNetworkBoundResource.kt

```diff
@@ -0,0 +1,152 @@
+    private val onFetchFailed: ((ApiErrorResponse<RequestType>) -> Unit)?
+) {
+    @ExperimentalCoroutinesApi
+    private val result = liveData<Resource<ResultType>> {
+        if (initialValue?.status != Status.SUCCESS) {
+            yield(Resource.loading(initialValue?.data))
+        }
+        val dbSource = loadFromDb()
+        val initialValue = dbSource.await()
+        val willFetch = initialValue == null || shouldFetch(initialValue)
+        if (!willFetch) {
+            // if we won't fetch, just yield existing db values as success
+            yieldSource(dbSource.map {
+                Resource.success(it)
+            })
+        } else {
+            doFetch(dbSource, this)
+        }
+    }
+
+    private suspend fun doFetch(
+        dbSource: LiveData<ResultType>,
+        liveDataScope: LiveDataScope<Resource<ResultType>>
+    ) {
+        // yield existing values as loading while we fetch
+        liveDataScope.yieldSource(dbSource.map {
+            Resource.loading(it)
+        })
+        val response = fetchCatching()
+        when (response) {
+            is ApiSuccessResponse, is ApiEmptyResponse -> {
+                if (response is ApiSuccessResponse) {
+                    val processed = processResponse(response)
+                    liveDataScope.clearSource()
+                    // before saving it, disconnect it so that new values comes w/ success
+                    saveCallResult(processed)
+                }
+                liveDataScope.yieldSource(loadFromDb().map {
+                    Resource.success(it)
+                })
+            }
+            is ApiErrorResponse -> {
+                onFetchFailed?.invoke(response)
+                liveDataScope.yieldSource(dbSource.map {
+                    Resource.error(response.errorMessage, it)
+                })
+            }
+        }
+    }
+
+    // temporary here during migration
+    fun asLiveData() = result as LiveData<Resource<ResultType>>
+
+    private suspend fun fetchCatching(): ApiResponse<RequestType> {
+        return try {
+            fetch()
+        } catch (ex: CancellationException) {
+            throw ex
+        } catch (ex: Throwable) {
+            ApiResponse.create(ex)
+        }
+    }
+
+    /**
+     * temporary workaround until we have cancellable yieldSource
+     */
+    private suspend fun <T> LiveDataScope<T>.clearSource() {
+        yieldSource(MutableLiveData<T>())
+    }
+
+    private suspend fun <T> LiveData<T>.await() = withContext(Dispatchers.Main) {
+        val receivedValue = CompletableDeferred<T?>()
+        val observer = Observer<T> {
+            if (receivedValue.isActive){
+                receivedValue.complete(it)
+            }
+        }
+        try {
+            observeForever(observer)
+            return@withContext receivedValue.await()
+        } finally {
+            removeObserver(observer)
```

Comment:
Does it work cool with cancellation?  I assume yes, because `receivedValue.await()` throws `CancelledException` once scope is cancelled and then will remove observer is it right?

Reply:
yes but i shall add a test for it.

Reply:
done.